components/locale/src/exemplar_chars.rs

```diff
@@ -0,0 +1,300 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module provides APIs for getting exemplar characters for a locale.
+//!
+//! Exemplars are characters used by a language, separated into different sets.
+//! The sets are: main, auxiliary, punctuation, numbers, and index.
+//!
+//! The sets define, according to typical usage in the language,
+//! which characters occur in which contexts with which frequency.
+//! For more information, see the documentation in the
+//! [Exemplars section in Unicode Technical Standard #35](https://unicode.org/reports/tr35/tr35-general.html#Exemplars)
+//! of the LDML specification.
+//!
+//! # Examples
+//!
+//! ```
+//! use icu::locale::locale;
+//! use icu::locale::exemplar_chars::ExemplarCharacters;
+//!
+//! let locale = locale!("en-001").into();
+//! let exemplars_main = ExemplarCharacters::try_new_main(&locale)
+//!     .expect("locale should be present");
+//!
+//! assert!(exemplars_main.contains_char('a'));
+//! assert!(exemplars_main.contains_char('z'));
+//! assert!(exemplars_main.contains("a"));
+//! assert!(!exemplars_main.contains("ä"));
+//! assert!(!exemplars_main.contains("ng"));
+//! ```
+
+use crate::provider::*;
+use core::ops::Deref;
+use icu_collections::codepointinvliststringlist::CodePointInversionListAndStringList;
+use icu_provider::prelude::*;
+
+/// A wrapper around `UnicodeSet` data (characters and strings)
+#[derive(Debug)]
+pub struct ExemplarCharacters {
+    data: DataPayload<ErasedUnicodeSetMarker>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub(crate) struct ErasedUnicodeSetMarker;
+impl DynamicDataMarker for ErasedUnicodeSetMarker {
+    type DataStruct = ExemplarCharactersV1<'static>;
+}
+
+impl ExemplarCharacters {
+    /// Construct a borrowed version of this type that can be queried.
+    ///
+    /// This avoids a potential small underlying cost per API call (ex: `contains()`) by consolidating it
+    /// up front.
+    #[inline]
+    pub fn as_borrowed(&self) -> ExemplarCharactersBorrowed<'_> {
+        ExemplarCharactersBorrowed {
+            data: self.data.get(),
+        }
+    }
+}
+
+/// A borrowed wrapper around code point set data, returned by
+/// [`ExemplarCharacters::as_borrowed()`]. More efficient to query.
+#[derive(Clone, Copy, Debug)]
+pub struct ExemplarCharactersBorrowed<'a> {
+    data: &'a ExemplarCharactersV1<'a>,
+}
+
+impl<'a> Deref for ExemplarCharactersBorrowed<'a> {
+    type Target = CodePointInversionListAndStringList<'a>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.data.0
+    }
+}
```

Comment:
Issue:

1. We might not want to restrict ourselves to returning `&CodePointInversionListAndStringList`, in case the data model changes. Counter-argument: it is the data model and we're not planning to change it.
2. `Deref` might be too strong a trait. ExemplarCharactersBorrowed adds more semantic meaning, in two levels: the data struct and the class itself. A function such as `.as_inversion_list()` might be better. Counter-argument: under the hood it is a wrapper so we should just impl Deref.

I lean toward doing the "safer" thing of just implementing `.contains` and friends on `ExemplarCharactersBorrowed` like was there before, but if you want to advocate for this other style, I can be convinced.

Reply:
Well the type is still `ExemplarCharactersBorrowed`, it just gains all the methods from `CodePointInversionListAndStringList`. I'd rather not copy-paste all its methods, that's why `Deref` exists.

Reply:
I understand that, _but you didn't address my concerns with `Deref`_.

Adding `Deref`, or any function that returns `&CodePointInversionListAndStringList`, locks us into a particular representation and behavior. Note that `CodePointSetData` has a function that returns `CodePointInversionList<'a>`, which is a `Cow`. This was a deliberate part of the design.

My position is that it is not very costly for us to add a conversion method and "copy-paste" the 2-4 key accessor methods, and it is a low enough cost that we should not lock ourselves into a representation.

I also make the further observation that it is usually a flawed design to return a reference-to-a-cow. You should instead be returning a fully borrowed type. So, if you did implement `Deref`, the correct thing would be to make `ExemplarCharactersBorrowed` wrap a new type called `CodePointInversionListAndStringListBorrowed` and then deref to that.

How about this: I could get behind a solution where `ExemplarCharacters` has a function that returns the cow type `CodePointInversionListAndStringList`, borrowing from the `ExemplarCharacters`, and we say that is the primary way to interface with it.

Reply:
> My position is that it is not very costly for us to add a conversion method and "copy-paste" the 2-4 key accessor methods, and it is a low enough cost that we should not lock ourselves into a representation.

I understood from a previous comment of yours that you're fine with locking us into a representation?

> I also make the further observation that it is usually a flawed design to return a reference-to-a-cow.

Where am I returning a reference to a cow?